### PR TITLE
CIP-0001 | Structure and template revisions

### PR DESCRIPTION
**Pull request by rphair**
A few months since @Ryun1 made the general improvement to versioning expectations in #563, and now in the course of #389, we are seeing Versioning treatments stand out as a common feature & therefore looking again at adding a Versioning subsection (of Specification) as an obligatory heading.

This was considered to add to the remediation effort since https://github.com/cardano-foundation/CIPs/issues/389#issuecomment-1660405709 and now @Ryun1 and I think it would be easier for people to check in general if there were a common subheading to review.

This addition & how it is worded should of course be unanimously agreed upon by all active CIP editors.  CIP authors & reviewers who are interested in the versioning issue should also please voice any opinions & suggestions here.

It also fixes some spelling and other chronic annoyances like the `/pulls` URI syntax error in the 2 templates.

---

([updated CIP-0001](https://github.com/rphair/CIPs/blob/add-versioning-section/CIP-0001/README.md))
([updated CIP template](https://github.com/rphair/CIPs/blob/add-versioning-section/.github/CIP-TEMPLATE.md))
([updated CPS template](https://github.com/rphair/CIPs/blob/add-versioning-section/.github/CPS-TEMPLATE.md))

### CODE REVIEW

**Comment by rphair** [comment]
last commit = another small update: CIP-0084 was still linking to the PR (which hadn't been merged yet at the time that CIP-0001 section was being rewritten)

File: .github/CPS-TEMPLATE.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ Authors:
     - John Doe <john.doe@email.domain>
 Proposed Solutions: []
 Discussions:
-    - https://github.com/cardano-foundation/cips/pulls/?
+    - https://github.com/cardano-foundation/cips/pull/?
```
Comment (Ryun1) [inline .github/CPS-TEMPLATE.md:10]:
```suggestion
    - https://github.com/cardano-foundation/CIPs/pull/?
```

File: CIP-0001/README.md [hunk 2]
```diff
@@ -249,7 +249,7 @@ Registered categories for explicitly enlisted projects are otherwise listed belo
 Category | Description
 ---      | ---
 Plutus   | Changes or additions to Plutus, following the process described in [CIP-0035][]
-Ledger   | For proposals regarding the Cardano ledger, following the process described in [CIP-0084][https://github.com/cardano-foundation/CIPs/pull/456]
+Ledger   | For proposals regarding the Cardano ledger, following the process described in [CIP-0084][]
```
Comment (Ryun1) [inline CIP-0001/README.md:252]: you will need to add the CIP-0084 link to the references (past line 442) :)

File: .github/CPS-TEMPLATE.md (hunk 1 above)
Comment (rphair) [reply .github/CPS-TEMPLATE.md:10]: grrr; I'd fixed one but forgot the other...

**Review by Ryun1 — COMMENTED** [review]
I think enforcing authors to be explicitly version their proposals makes a lot of sense.
The CIPs yearn to be versioned!! 😆

Although I suggested we add a subsection to the `Specification` I think this alone may cause problems:
- How should we expect authors to version the other sections of their proposal? is this even needed?
- Should we just rely on git history, to allow looking at past versions? or store different versions together?
- Should authors have full freedom to version their proposal? or is there a limit to where a new proposal should be written?
- Where can rationale for changes and version bumps be found?

File: CIP-0001/README.md (hunk 2 above)
Comment (rphair) [reply CIP-0001/README.md:252]: fixed as of https://github.com/cardano-foundation/CIPs/pull/730/commits/0708c7e5905c96105f2a6de216519eafe076f930

Comment (Ryun1) [reply CIP-0001/README.md:252]: :)

**Comment by rphair** [comment]
As I assumed... which, if uncertain, may need to be said explicitly... the `Versioning` is part of the `Specification` but applies to the whole proposal, and therefore all of its sections.  It's versioning of the _proposal_, not of the document.

If people are looking for document versioning, that's where GitHub comes in.  It would be inadequate as well as redundant (a duplication of labour, already automatically performed by the UI) to document text changes in the document itself.

I also believe versioning the text, rather than the functional spec, would quickly lead to unwieldly documents that authors won't want to help maintain and that would be a burden to keep up to date for a small CIP editing team.

If we can agree (and if so I'll add to CIP-0001) that CIP text is versioned on GitHub and the CIP behaviour is versioned in the new `Versioning` section, then everything else except that clarification works as presented and we can continue moving forward more-or-less as we were + the incremental _functional_ versioning already agreed to by the community.

> Should authors have full freedom to version their proposal? or is there a limit to where a new proposal should be written?

We would have to consider what's in the `Versioning` section of equal rank as anything else in a CIP.  So if an author presented an incomplete or unreasonable versioning plan, like everything else it would be fixed in the review process between authors, editors, and community input.

> Where can rationale for changes and version bumps be found?

I would propose adding to the new Versioning section requirements that Rationale for version bumps should be included as they normally would be summarised in familiar software "Release notes" which are sometimes as simple as one explanatory line about what the changes mainly accomplish.
- This is another matter in which the level of detail should be accepted based on review with editors and community advocates: if too much or too little for a particular CIP, that will be revealed in the editorial process.

**Comment by rphair** [comment]
@michaelpj your input would also be welcome, since you were so involved in our last big rewrite of CIP-0001 when structural headings & also some aspects of versioning were considered.

File: .github/CIP-TEMPLATE.md [hunk 3]
```diff
@@ -7,7 +7,7 @@ Authors:
     - John Doe <john.doe@email.domain>
 Implementors: []
 Discussions:
-    - https://github.com/cardano-foundation/cips/pulls/?
+    - https://github.com/cardano-foundation/CIPs/pull/?
```
Comment (Ryun1) [inline .github/CIP-TEMPLATE.md:10]:
```suggestion
  - https://github.com/cardano-foundation/CIPs/pull/?
```
should these be 2 or 4 spaces? (same for author too)

File: .github/CIP-TEMPLATE.md [hunk 4]
```diff
@@ -46,7 +49,7 @@ It must also explain how the proposal affects the backward compatibility of exis
 <!-- Describes what are the acceptance criteria whereby a proposal becomes 'Active' -->
 
 ### Implementation Plan
-<!-- A plan to meet those criteria. Or `N/A` if not applicable. -->
+<!-- A plan to meet those criteria or `N/A` if an implementation plan is not applicable. -->
 
 ## Copyright
 <!-- The CIP must be explicitly licensed under acceptable copyright terms. -->
```
Comment (Ryun1) [inline .github/CIP-TEMPLATE.md:55]: should we add?:

`This CIP is licensed under [CC-BY-4.0][].`

Would make it slightly easier for authors and also ensures the `.` 😆.

File: .github/CIP-TEMPLATE.md (hunk 3 above)
Comment (rphair) [reply .github/CIP-TEMPLATE.md:10]: @Ryun1 technically it doesn't matter.  Generally we see 2, 4, or 8 spaces... the only things that matter to list formatting are when the number of spaces increase (which indents) and when they line up with a previous number of spaces (which outdents).

Personally I've always stuck with the standard first posted by the document author, unless indenting an outline structure for the first time: in which case I use a 2-space increment.

A bunch of @KtorZ authored CIPs have the 4-space standard which is fine with me as well.  I don't think we'd be saving ourselves any grief if we tried to mandate or standardise on a particular spacing increment. 🤔

File: .github/CIP-TEMPLATE.md (hunk 4 above)
Comment (rphair) [reply .github/CIP-TEMPLATE.md:55]: Technically yes, but at best it would violate the principle of a template if they intended to use the Apache license, and at worst it would have a different license in the header than in the footer.

In fact I believe the several Apache vs. CC license mismatches we have in remediated (#389) CIPs are from the time the license markup _was_ in an earlier version of the template.

Comment (Ryun1) [reply .github/CIP-TEMPLATE.md:55]: good spot, we shouldnt then!

I am just trying to make it slightly easier for authors and ourselves, as often the `Copyright` section is neglected to be filled in .

Comment (Ryun1) [reply .github/CIP-TEMPLATE.md:55]: > it would violate the principle of a template if they intended to use the Apache license

Well the template includes `CC-BY-4.0` in the header by default 😆

What about, in the `Copyright` section:

```
<!-- This CIP is licensed under [CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/legalcode). -->
<!-- This CIP is licensed under [Apache-2.0](http://www.apache.org/licenses/LICENSE-2.0). -->
```

and we remove the potentially confusing markdown references part at the bottom?

File: .github/CIP-TEMPLATE.md (hunk 3 above)
Comment (Ryun1) [reply .github/CIP-TEMPLATE.md:10]: > technically it doesn't matter

ahh good, I hadnt found any examples of it breaking things, but was unsure.

>  don't think we'd be saving ourselves any grief

Agreed, it was just something that has bugged me and I wanted to raise. 
although it is of course very insignificant

File: .github/CIP-TEMPLATE.md (hunk 4 above)
Comment (rphair) [reply .github/CIP-TEMPLATE.md]: sure; let's do it that way (added some instruction as well, to try to reduce even more repetitive editing): https://github.com/cardano-foundation/CIPs/pull/730/commits/61974c20b3cac8c90837a4244c47591c4fc38def

I am also now wondering why the `Copyright` field or section aren't in the CPS template.  I was around for that big remodel but don't recall seeing why at least CC licensing isn't recommended since even the simplest license have a standard open source liability waiver that might protect interested parties in the case of some ambitious CPS. (cc @KtorZ @michaelpj)

**Comment by michaelpj** [comment]
It seems to me that there are multiple kinds of versioning we might care about:
- A CIP can be versioned as a document _itself_. This is in principle handled by git, but that information is not easily visible by, say, someone reading the rendered site. It's common for text documents that are re-presented in this way to have a section that summarises changes to the document.
    - e.g. https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#changeLog
- A CIP may _define_ one (or many!) specifications which may require versioning. This is different because the _current_ version of the CIP _document_ might need to clearly explain _all+ different versions of the specification if, say, they are all in use at one time.
    - e.g. the Plutus Core specification specifies all language variants in one document (although it _doesn't_ summarise its document history!)

So I imagine something like this:
```
# Abstract

Defines the frobnicate and quxify file specifications.

# Specification

## Frobnicate

### Version 1

This defines the specification for the original Version 1 of the frobnicate specification ...

### Version 2

This defines the specification for Version 2 of the frobnicate specification, which differs from Version 1 in the following ways ...

## Quxify

### Version 1

...

### Version 2

....

### Version 3

...

# Changelog

## <date1>

Initial version of the CIP, adding version 1 of the frobnicate and quxify specs

## <date2>

Significant re-wording of the frobnicate spec to clear up ambiguities. The new version is authoritative.

## <date3>

Added version 2 of the quxify spec

## <date3>

Added version 2 of the frobnicate spec and version 3 of the quxify spec
```

**Comment by rphair** [comment]
@michaelpj OK, so this proposing two points different form the last things I said here:
1. `Versioning` can apply to components of the _Specification_ as well as the whole spec.
2. `Changelog` should be included in CIPs rather than relying to GitHub to produce it dynamically.

I agree with point 1 now that you bring it up.  A good case in point might have been CIP-0068 if the versioning weren't already included as part of the data structure itself.  It's more future proof to have `Versioning` apply to any subsection of the specification... I'll propose a rewrite as soon as we have more opinions in this thread and #731 (we still have holiday attrition and then I may need to take a break myself).

Point 2 at first seems like a nice framework to explain how CIP text evolves: until you see how people actually use it.  CIP authors will generally post background material for updates like they post information on GitHub releases: i.e. _nothing_.

With more examples to go on in the early days, authors felt compelled to produce `Changelog`s like this; showing how a text _Changelog_ can actually make things worse, rather than better, when a document is edited by committee:

- https://github.com/cardano-foundation/CIPs/blob/0a487d9e482944642215af7328013a4f6f32a592/CIP-0027/README.md#change-log

For brevity's sake we would probably ask them to summarise issues in one line: but:
- this is already being done by PR title and sometimes individual commit titles in the GitHub file and/or directory history.
- on that note, the directory history will interleave changes to subordinate files, which we can't rely on authors/editors to maintain in the markup.
- And if we're expecting more descriptive changelogs, does the reader really benefit from having a CIP author and/or editor summarise what can be cleaned from spending a few minutes reading each applicable PR?

Also we've been following the standard of putting substantial update PRs in the `Discussions:` link in the header.  I believe this is more useful than a Changelog because:
- it's machine readable and can produce hypertext maps when CIP text is incorporated into other types of derived web sites / databases
- it's already filtered by the CIP author and/or editor(s) to remove text updates that don't change anything substantial (i.e. typos, formatting, updated URLs, administrative features).

**TL;DR** I would therefore say `yes` to the recursive application of `Versioning` but `no` for the imposition of an error-prone document `Changelog` 😜

**Comment by michaelpj** [comment]
> It's more future proof to have Versioning apply to any subsection of the specification... 

I think it's hard to say something more structured than what you already have: a statement that the Specifciation should explain any versioning considerations.

Because versioning is _wildly heterogeneous_. Depending what the CIP is specifying, what kind of versioning makes sense will be quite different. So I think the simplest thing to do is just to make it clear that you're supposed to talk about it.

> Point 2 at first seems like a nice framework to explain how CIP text evolves: until you see how people actually use it. CIP authors will generally post background material for updates like they post information on GitHub releases: i.e. nothing.

That's fair, I do agree that getting people to write text changelogs is hard.

> Also we've been following the standard of putting substantial update PRs in the Discussions: link in the header. I believe this is more useful than a Changelog because:

That's pretty reasonable. I have a couple of suggestions:
- Add an actual metadata section for it instead of using `Discussions`. You want it to be parseable, so give it a proper name!
- When rendering it, display the _name_ of the PR also. That will hopefully give an approximate idea of what changed in each one.
- You could consider actually generating that from the git history if you wanted to when you generate the site. Unsure if that's better or worse.

**Comment by rphair** [comment]
In light of @michaelpj https://github.com/cardano-foundation/CIPs/pull/730#issuecomment-1875714369 (elaborating on @Ryun1 https://github.com/cardano-foundation/CIPs/pull/730#pullrequestreview-1799847551) I've removed the `### Versioning` section that Ryan & I had originally considered in PM immediately before I posted it in this PR.

Since as per @KtorZ https://github.com/cardano-foundation/CIPs/issues/731#issuecomment-1876820203 we could still clarify use of some optional sections, we can still iffer a top level `## Versioning` section in case authors want to use it to characterise their entire versioning approach and to ask them not to use it as a Changelog for the text itself.

I've adjusted this with a heading for optional sections that have come up so far in both the issue & PR, and moved the Versioning requirements into a subsection where they can be more easily maintained.

**Review by Crypto2099 — APPROVED** [review]
Looks good to me, can't think of anything I would currently add or modify at this point. I think the versioning requirement is sufficiently broad to allow authors to determine a versioning strategy as they see fit while still hopefully encouraging the "best practice" of ensuring that authors are thinking about these things during their writing.

**Comment by rphair** [comment]
Also now I've ticked the final 2 boxes in the Implementation plan now that the #389 remediation is complete.

@KtorZ there is a further comment about notifying enlisted project maintainers of projects categorised to them, now ticked off as a side effect... but I can't think of any CIP for which this hasn't already happened naturally somehow over the last year or so.

**Review by Ryun1 — APPROVED** [review]
I think this is a really great step forward.

**Comment by rphair** [comment]
thanks; will merge pre-meeting then since still a lot of PRs coming in with some really live ones up for review again, and this is mainly a matter for editor deliberation that the other usual attendees won't feel strongly about.
